ENH: sosfreqz

Eric89GXL opened the pull request:
I had no idea how close to mergeable @WarrenWeckesser's #4465 was. This should be ready for review/merge, I just rebased and tweaked the release notes.

e-q commented:
Thanks for picking this up! I got sucked down a rabbit hole when trying to make this part of gh-6059, which I will now ramble on about:

Basically, I was thinking it would be good if proper SOS and ZPK frequency responses could just live behind `(d)freqresp` without needing to create more public functions. This led me to look into creating the SOS LTI objects classes which we definitely want; however, this breaks all of the functions with call signatures that count arguments to determine the filter representation. 

I.e., we specify the SOS arrays as arrays of shape `(N, 6)`, but functions like `freqresp` use their arguments to create an LTI object via `lti(*args)` which means `lti(*sos)` can look like a TF, zpk, or SS depending on `N`. So, we need to either deprecate those call signatures, or find a way to determine the difference between a `(2,6)` SOS array and length six `(b, a)` coefficients.

All this is to say: we probably do indeed need this, and the more accurate `zpk` version, to be public. We should also come up with a solution to the SOS LTI object issue sooner rather than later.

Eric89GXL commented:
> we probably do indeed need this, and the more accurate zpk version, to be public

Maybe just for now we do the SOS, and then someone else can make public the improved ZPK one?

> We should also come up with a solution to the SOS LTI object issue sooner rather than later.

Yeah I would like to see that, too, but I worry about holding this one up to wait on it. Are you okay going forward with `sosfreqz` for now?

e-q commented:
> Yeah I would like to see that, too, but I worry about holding this one up to wait on it. Are you okay going forward with sosfreqz for now?

Sorry for being unclear, this is what I was getting at. I agree that this PR should happen sooner rather than later.

WarrenWeckesser commented on doc/release/0.19.0-notes.rst:
```diff
@@ -22,6 +22,11 @@ This release requires Python 2.7 or 3.4-3.5 and NumPy 1.7.1 or greater.
+`scipy.signal` improvements
+---------------------------
+
+The funtion `scipy.signal.sosfreqz` was added to compute the frequency
```
function

WarrenWeckesser commented on scipy/signal/filter_design.py:
```diff
@@ -353,6 +354,127 @@ def group_delay(system, w=None, whole=False):
+
+    """
+
+    # A valid SOS array must have a least one section, so it should be
```
This comment is no longer relevant.

WarrenWeckesser commented on scipy/signal/filter_design.py:
```diff
@@ -353,6 +354,127 @@ def group_delay(system, w=None, whole=False):
+    sos, n_sections = _validate_sos(sos)
+    if n_sections == 0:
+        raise ValueError('Cannot compute frequencies with no sections')
+    w = h = 1.
```
There is no need to initialize `w`.

WarrenWeckesser commented:
Eric, thanks for picking this up.  The changes look good; I made a few suggestions the code.

My PR was pretty close to ready.  The main issue (which I agree with) was the API.  I copied it from `freqz` just to be consistent, but it isn't a great starting point.  (`worN` is an awkward argument name, and if not for maintaining backwards compatibility, I'd remove the `plot` argument from `freqz`.)  Alternatives have been discussed elsewhere, but their implementation is not imminent, so perhaps it would be best to get `sosfreqz` in now, knowing that it might become obsolete with the next generation of the signal API.  The perfect is the enemy of the good and all that...

Eric89GXL commented:
@WarrenWeckesser any idea how your changes could have broken some `scipy.special` tests for `mpmath`?

Eric89GXL commented:
I removed the `plot` argument since I don't think it should really be there (even in `freqz`), but I kept the other arguments named the same so there is some compatibility with the existing function.

person142 commented on scipy/signal/tests/test_filter_design.py:
```diff
@@ -424,6 +441,51 @@ def plot(w, h):
+        N = 500
+        order = 25
+        Wn = 0.15
+        mpmath.mp.dps = 80
```
@Eric89GXL: I think the test failure comes from increasing the dps here. Don't understand how the increased precision persists across the tests, but if you run ``test_hyp0f1_gh5764`` with a dps of 80 it fails. IOW ``test_hyp0f1_gh5764`` is bad... but at low precision mpmath happens to be bad in the same way.

I bet restoring the original precision at the end of ``test_sos_freqz_against_mp`` will fix the problem.

Eric89GXL replied:
Ahh I tried to find some global setting but missed this. Thanks for finding it, I'll set it back

Eric89GXL commented:
@endolith do you want to have a look since you have been involved in SOS?

@rgommers you've been involved with the API discussions, can you live with another function that isn't part of the grand-unified API?

person142 commented on scipy/signal/tests/test_filter_design.py:
```diff
@@ -424,6 +441,55 @@ def plot(w, h):
+        Wn = 0.15
+        old_dps = mpmath.mp.dps
+        mpmath.mp.dps = 80
+        try:
```
FWIW the ``mpmath.workdps`` context manager is nice for doing this sort of thing.

Eric89GXL replied:
Cool, I'll put that in

rgommers commented:
> @rgommers you've been involved with the API discussions, can you live with another function that isn't part of the grand-unified API?

Yes. One more or less is not going to make the difference, and this is functionality that is needed.

endolith commented on scipy/signal/filter_design.py:
```diff
@@ -353,6 +358,117 @@ def group_delay(system, w=None, whole=False):
+    a digital filter, compute the frequency response of the system function::
+
+               B0(z)*B1(z)*...*B{n-1}(z)
+        H(z) = -------------------------
```
How about 

```
               B0(z)   B1(z)         B{n-1}(z)
        H(z) = ----- * ----- * ... * ---------
               A0(z)   A1(z)         A{n-1}(z)
```

to imply the stages being calculated separately and then combined?

endolith commented on scipy/signal/filter_design.py:
```diff
@@ -353,6 +358,117 @@ def group_delay(system, w=None, whole=False):
+    if m != 6:
+        raise ValueError('sos array must be shape (n_sections, 6)')
+    if not (sos[:, 3] == 1).all():
+        raise ValueError('sos[:, 3] should be all ones')
```
Do these really need to be all ones?

Eric89GXL replied:
We should be guaranteed this (gains should be elsewhere) from the SciPy functions even though user-defined ones might not have it. But if we're going to change that we need to do it elsewhere and discuss it since this is in 0.18 and all I did was move it.

endolith commented on scipy/signal/filter_design.py:
```diff
@@ -353,6 +358,117 @@ def group_delay(system, w=None, whole=False):
+
+    Notes
+    -----
+    Using Matplotlib's "plot" function as the callable for `plot` produces
```
this can be removed since plot was removed

endolith commented on scipy/signal/filter_design.py:
```diff
@@ -353,6 +358,117 @@ def group_delay(system, w=None, whole=False):
+    Returns
+    -------
+    w : ndarray
+        The normalized frequencies at which h was computed, in radians/sample.
```
variables should be in single backticks:

    The normalized frequencies at which `h` was computed

endolith commented on scipy/signal/filter_design.py:
```diff
@@ -353,6 +358,117 @@ def group_delay(system, w=None, whole=False):
+    Examples
+    --------
+
+    >>> from scipy import signal
```
I think this should be combined with the `sos = ` line

endolith commented on scipy/signal/filter_design.py:
```diff
@@ -353,6 +358,117 @@ def group_delay(system, w=None, whole=False):
+    --------
+
+    >>> from scipy import signal
+    >>> import matplotlib.pyplot as plt
```
and this should be combined with the `plt.subplot...` line

endolith commented:
Also would it be worth showing a higher-order filter that becomes garbled with freqz?

endolith commented on scipy/signal/filter_design.py:
```diff
@@ -353,6 +358,117 @@ def group_delay(system, w=None, whole=False):
+    >>> sos = signal.ellip(8, 0.5, 60, (0.2, 0.4), btype='bandpass',
+    ...                    output='sos')
+
+    Compute the frequency response on a grid of 1500 points.
```
`Compute the frequency response at 1500 points from DC to Nyquist.`?

endolith commented on scipy/signal/filter_design.py:
```diff
@@ -353,6 +358,117 @@ def group_delay(system, w=None, whole=False):
+    >>> plt.yticks([-np.pi, -0.5*np.pi, 0, 0.5*np.pi, np.pi],
+    ...            ['$-\pi$', '$-\pi/2$', '0', '$\pi/2$', '$\pi$'])
+    >>> plt.ylabel('Phase [rad]')
+    >>> plt.xlabel('Frequency (relative to Nyquist)')
```
"Normalized frequency"?

endolith commented on scipy/signal/filter_design.py:
```diff
@@ -353,6 +358,117 @@ def group_delay(system, w=None, whole=False):
+    w : ndarray
+        The normalized frequencies at which h was computed, in radians/sample.
+    h : ndarray
+        The frequency response.
```
`The frequency response, as complex numbers.` to clarify it's not just amplitude?

endolith commented:
Also would it make sense to have a parameter that makes it output the responses for each stage separately, so you can plot them individually?  https://en.wikibooks.org/wiki/Signals_and_Systems/Second_Order_Transfer_Function#Amplitude_Responses

endolith commented on scipy/signal/filter_design.py:
```diff
@@ -353,6 +358,117 @@ def group_delay(system, w=None, whole=False):
+
+    sos, n_sections = _validate_sos(sos)
+    if n_sections == 0:
+        raise ValueError('Cannot compute frequencies with no sections')
```
Should `freqz([], [], worN=10)` raise a `ValueError`, too?

Eric89GXL replied:
Perhaps, but it might be a backward incompatible change...?

endolith commented on scipy/signal/filter_design.py:
```diff
@@ -353,6 +358,117 @@ def group_delay(system, w=None, whole=False):
+    >>> plt.ylabel('Gain [dB]')
+    >>> plt.title('Frequency Response')
+    >>> plt.subplot(2, 1, 2)
+    >>> plt.plot(w/np.pi, np.arctan2(h.imag, h.real))
```
`np.angle(h)` is easier to read

endolith commented:
Suggested changes in https://github.com/endolith/scipy/commit/b35fbc63b3e67483cd0a05a44afe8332b534886c

Eric89GXL commented:
Thanks @endolith cherry-picked and updated.

Based on the comments I think this should be mergeable. If anyone else wants to look again (or wants me to wait) feel free, otherwise I'll put it in next week.

endolith commented:
For testing, you could generate some high-order filters as SOS and then check that sosfreqz is within bounds for what the filter should do, like the plots and tests for these:

- http://docs.scipy.org/doc/scipy-0.14.0/reference/generated/scipy.signal.cheb2ord.html 
- http://docs.scipy.org/doc/scipy-0.14.0/reference/generated/scipy.signal.cheb1ord.html 
- http://docs.scipy.org/doc/scipy-0.14.0/reference/generated/scipy.signal.ellipord.html 
- http://docs.scipy.org/doc/scipy-0.14.0/reference/generated/scipy.signal.buttord.html
- https://github.com/scipy/scipy/blob/master/scipy/signal/tests/test_filter_design.py#L556-896

Eric89GXL commented:
@endolith commit added, see if you're happy now

endolith commented:
I started modifying the test cases to make them more extreme (so that sosfreqz could handle them but freqz couldn't) but one of them fails, and I don't now if it's `sosfreqz` or `ellipord` or `ellip` that's actually responsible:

```
N, Wn = ellipord(0.3, 0.25, .5, 150)
sos = ellip(N, .5, 150, Wn, 'high', output='sos')
w, h = sosfreqz(sos)
dB = 20*np.log10(np.abs(h))
w /= np.pi
plot(w, dB)
```

![shouldn't go over -150 db](https://cloud.githubusercontent.com/assets/58611/16750592/de17c59c-47a0-11e6-8ad6-b2eb7020f9ed.png)

WarrenWeckesser commented:
@endolith wrote:

> I started modifying the test cases to make them more extreme (so that sosfreqz could handle them but freqz couldn't) 

That's also what `TestSOSFreqz.test_sos_freqz_against_mp()` does--it verifies the correctness of an order 25 filter by comparing it to the result computed using `mpmath`.  The `mpmath` functions are in `tests/mpsig.py`.

endolith commented:
anyway these are the checks I made:

```
        # from cheb2ord
        N, Wn = cheb2ord([0.1, 0.6], [0.2, 0.5], 3, 150)
        sos = cheby2(N, 150, Wn, 'stop', output='sos')
        w, h = sosfreqz(sos)
        dB = 20*np.log10(np.abs(h))
        w /= np.pi
        assert_allclose(dB[w <= 0.1], 0, atol=3.01)
        assert_allclose(dB[w >= 0.6], 0., atol=3.01)
        assert_array_less(dB[(w >= 0.2) & (w <= 0.5)], -149.9)

        # from cheb1ord
        N, Wn = cheb1ord(0.2, 0.3, 1, 150)
        sos = cheby1(N, 1, Wn, 'low', output='sos')
        w, h = sosfreqz(sos)
        dB = 20*np.log10(np.abs(h))
        w /= np.pi
        assert_allclose(dB[w <= 0.2], 0, atol=1.01)
        assert_array_less(dB[w >= 0.3], -149.9)

        # adapted from ellipord
        N, Wn = ellipord(0.3, 0.25, .5, 150)
        sos = ellip(N, .5, 150, Wn, 'high', output='sos')
        w, h = sosfreqz(sos)
        dB = 20*np.log10(np.abs(h))
        w /= np.pi
        assert_allclose(dB[w >= 0.3], 0, atol=.55)
        assert_array_less(dB[(w > 0) & (w <= 0.25)], -149.9)

        # adapted from buttord
        N, Wn = buttord([0.2, 0.5], [0.14, 0.6], 3, 100)
        sos = butter(N, Wn, 'band', output='sos')
        w, h = sosfreqz(sos)
        dB = 20*np.log10(np.abs(h))
        w /= np.pi
        assert_array_less(dB[(w > 0) & (w <= 0.14)], -99.9)
        assert_array_less(dB[w >= 0.6], -99.9)
        assert_allclose(dB[(w >= 0.2) & (w <= 0.5)], 0, atol=3.01)
```

Eric89GXL commented:
Thanks @endolith added your tests (after making the problematic `ellipord` test use 147 instead of ~149 for the acceptable level), squashed and rebased. 

If there are no other comments I'll merge in a couple of days.

Eric89GXL commented:
Thanks for the reviews
